File: src/cmd/run_watch.zig
//! M22_001 §1: SSE streaming for `zombied run --watch`.
//!
//! Connects to GET /v1/runs/{id}:stream, buffers the SSE response,
//! then parses and renders gate_result events to stdout.

const std = @import("std");

/// SSE line parser state. Buffers bytes until '\n', dispatches events on blank lines.
pub const SseState = struct {
    alloc: std.mem.Allocator,
    line_buf: std.ArrayList(u8),
    current_data: ?[]u8,
    current_event: ?[]u8,
    done: bool,

    pub fn init(alloc: std.mem.Allocator) SseState {
        return .{
            .alloc = alloc,
            .line_buf = .{},
            .current_data = null,
            .current_event = null,
            .done = false,
        };
    }

    pub fn deinit(self: *SseState) void {
        self.line_buf.deinit(self.alloc);
        if (self.current_data) |d| self.alloc.free(d);
        if (self.current_event) |e| self.alloc.free(e);
    }

    fn processLine(self: *SseState, raw: []const u8) void {
        const trimmed = std.mem.trimRight(u8, raw, "\r");
        if (trimmed.len == 0) {
            if (self.current_data) |data| {
                const event_type = self.current_event orelse "message";
                renderSseEvent(self.alloc, event_type, data);
                if (std.mem.eql(u8, event_type, "run_complete")) self.done = true;
            }
            if (self.current_data) |d| self.alloc.free(d);
            self.current_data = null;
            if (self.current_event) |e| self.alloc.free(e);
            self.current_event = null;
            return;
        }
        if (std.mem.startsWith(u8, trimmed, "data: ")) {
            if (self.current_data) |d| self.alloc.free(d);
            self.current_data = self.alloc.dupe(u8, trimmed["data: ".len..]) catch null;
        } else if (std.mem.startsWith(u8, trimmed, "event: ")) {
            if (self.current_event) |e| self.alloc.free(e);
            self.current_event = self.alloc.dupe(u8, trimmed["event: ".len..]) catch null;
        }
    }

    pub fn feedBytes(self: *SseState, bytes: []const u8) void {
        for (bytes) |b| {
            if (b == '\n') {
                self.processLine(self.line_buf.items);
                self.line_buf.clearRetainingCapacity();
            } else {
                self.line_buf.append(self.alloc, b) catch {};
            }
        }
    }
};

pub fn streamRunOutput(
    alloc: std.mem.Allocator,
    base_url: []const u8,
    api_key: []const u8,
    run_id: []const u8,
) !void {
    const url = try std.fmt.allocPrint(alloc, "{s}/v1/runs/{s}:stream", .{ base_url, run_id });
    defer alloc.free(url);

    const auth_header = try std.fmt.allocPrint(alloc, "Bearer {s}", .{api_key});
    defer alloc.free(auth_header);

    var client = std.http.Client{ .allocator = alloc };
    defer client.deinit();

    const uri = try std.Uri.parse(url);

    // §1: Use client.open() for true incremental streaming.
    // client.fetch() buffers the entire response before returning — defeats SSE real-time.
    // client.open() + req.reader().read() delivers bytes as they arrive from the server.
    var header_buf: [4096]u8 = undefined;
    var req = try client.open(.GET, uri, .{
        .extra_headers = &.{
            .{ .name = "Authorization", .value = auth_header },
            .{ .name = "Accept", .value = "text/event-stream" },
        },
        .server_header_buffer = &header_buf,
    });
    defer req.deinit();

    try req.send();
    try req.wait();

    if (req.status != .ok) {
        std.debug.print("error: stream returned {d}\n", .{@intFromEnum(req.status)});
        return error.StreamFailed;
    }

    var sse = SseState.init(alloc);
    defer sse.deinit();
    var buf: [4096]u8 = undefined;
    while (true) {
        const n = req.reader().read(&buf) catch |err| {
            std.debug.print("error: stream read failed: {s}\n", .{@errorName(err)});
            break;
        };
        if (n == 0) break;
        sse.feedBytes(buf[0..n]);
        if (sse.done) break;
    }
}

pub fn renderSseEvent(alloc: std.mem.Allocator, event_type: []const u8, data: []const u8) void {
    if (std.mem.eql(u8, event_type, "run_complete")) {
        std.debug.print("[done] run complete: {s}\n", .{data});
        return;
    }
    if (!std.mem.eql(u8, event_type, "gate_result")) return;

    const parsed = std.json.parseFromSlice(std.json.Value, alloc, data, .{}) catch return;
    defer parsed.deinit();

    const obj = switch (parsed.value) {
        .object => |o| o,
        else => return,
    };

    const gate_name = if (obj.get("gate_name")) |v| switch (v) {
        .string => |s| s,
        else => "?",
    } else "?";
    const outcome = if (obj.get("outcome")) |v| switch (v) {
        .string => |s| s,
        else => "?",
    } else "?";
    const loop_n = if (obj.get("loop")) |v| switch (v) {
        .integer => |i| i,
        else => 0,
    } else 0;
    const wall_ms = if (obj.get("wall_ms")) |v| switch (v) {
        .integer => |i| i,
        else => 0,
    } else 0;

    std.debug.print("[{s}] {s} (loop {d}, {d}ms)\n", .{ gate_name, outcome, loop_n, wall_ms });
}

File: zombiectl/src/commands/run_watch.js
// M22_001 §5: --watch SSE streaming for zombiectl run command.

const WATCH_MAX_RETRIES = 3;
const WATCH_RETRY_DELAY_MS = 2000;
const WATCH_CONNECT_TIMEOUT_MS = 30_000;

/**
 * Stream SSE events for a run in real time.
 *
 * DX contract:
 *   - Ctrl+C / SIGINT cancels the stream cleanly (exit 0, no orphaned connection)
 *   - Network errors retry up to WATCH_MAX_RETRIES with Last-Event-ID for gap-free replay
 *   - Progress: "waiting for events..." printed when no event arrives within 30s
 *   - Connection timeout after WATCH_CONNECT_TIMEOUT_MS
 */
async function streamRunWatch(ctx, runId, { apiHeaders: getHeaders, ui, writeLine }) {
  const url = `${ctx.apiUrl}/v1/runs/${encodeURIComponent(runId)}:stream`;
  const fetchFn = ctx.fetchImpl || globalThis.fetch;
  let lastEventId = null;
  let aborted = false;

  const abortController = new AbortController();
  const onSigint = () => {
    aborted = true;
    abortController.abort();
  };
  process.on("SIGINT", onSigint);

  writeLine(ctx.stderr, ui.dim("watch: streaming (Ctrl+C to stop)"));

  try {
    for (let attempt = 0; attempt <= WATCH_MAX_RETRIES; attempt++) {
      if (aborted) break;

      const reqHeaders = { ...getHeaders(ctx), Accept: "text/event-stream" };
      if (lastEventId) reqHeaders["Last-Event-ID"] = lastEventId;

      let response;
      try {
        response = await fetchFn(url, {
          method: "GET",
          headers: reqHeaders,
          signal: abortController.signal,
        });
      } catch (err) {
        if (aborted || err.name === "AbortError") break;
        if (attempt < WATCH_MAX_RETRIES) {
          writeLine(ctx.stderr, ui.dim(`watch: connection failed, retrying (${attempt + 1}/${WATCH_MAX_RETRIES})...`));
          await new Promise((r) => setTimeout(r, WATCH_RETRY_DELAY_MS));
          continue;
        }
        writeLine(ctx.stderr, ui.err(`watch: connection failed after ${WATCH_MAX_RETRIES} retries: ${err.message}`));
        return;
      }

      if (!response.ok) {
        writeLine(ctx.stderr, ui.err(`watch: stream returned ${response.status}`));
        if (response.status >= 500 && attempt < WATCH_MAX_RETRIES) {
          await new Promise((r) => setTimeout(r, WATCH_RETRY_DELAY_MS));
          continue;
        }
        return;
      }

      const result = await readSseStream(response, ctx, lastEventId, {
        ui, writeLine,
      });
      lastEventId = result.lastEventId;

      if (result.done || aborted) break;

      if (result.error && attempt < WATCH_MAX_RETRIES) {
        writeLine(ctx.stderr, ui.dim(`watch: stream interrupted, reconnecting (${attempt + 1}/${WATCH_MAX_RETRIES})...`));
        await new Promise((r) => setTimeout(r, WATCH_RETRY_DELAY_MS));
        continue;
      }

      if (result.error) {
        writeLine(ctx.stderr, ui.err(`watch: stream error: ${result.error.message}`));
      }
      break;
    }
  } finally {
    process.removeListener("SIGINT", onSigint);
  }
}

async function readSseStream(response, ctx, lastEventIdIn, { ui, writeLine }) {
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let currentEvent = "message";
  let currentData = null;
  let lastActivity = Date.now();
  let heartbeatTimer = null;
  let lastEventId = lastEventIdIn;
  let streamDone = false;
  let streamError = null;

  const startHeartbeatCheck = () => {
    clearTimeout(heartbeatTimer);
    heartbeatTimer = setTimeout(() => {
      const idle = Math.round((Date.now() - lastActivity) / 1000);
      writeLine(ctx.stderr, ui.dim(`watch: waiting for events... (${idle}s idle)`));
      startHeartbeatCheck();
    }, WATCH_CONNECT_TIMEOUT_MS);
  };
  startHeartbeatCheck();

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      lastActivity = Date.now();
      startHeartbeatCheck();

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop();

      for (const raw of lines) {
        const line = raw.replace(/\r$/, "");
        if (line === "") {
          if (currentData !== null) {
            if (currentEvent === "run_complete") {
              writeLine(ctx.stdout, ui.ok(`run complete: ${currentData}`));
              streamDone = true;
              break;
            }
            if (currentEvent === "gate_result") {
              try {
                const d = JSON.parse(currentData);
                writeLine(ctx.stdout, `[${d.gate_name}] ${d.outcome} (loop ${d.loop}, ${d.wall_ms}ms)`);
              } catch { /* ignore malformed */ }
            }
            // M21_001 §1.4: interrupt acknowledgement from server
            if (currentEvent === "interrupt_ack") {
              try {
                const d = JSON.parse(currentData);
                writeLine(ctx.stdout, ui.dim(`→ steered (mode: ${d.mode})`));
              } catch { /* ignore malformed */ }
            }
          }
          currentEvent = "message";
          currentData = null;
          continue;
        }
        if (line.startsWith("data: ")) currentData = line.slice(6);
        else if (line.startsWith("event: ")) currentEvent = line.slice(7);
        else if (line.startsWith("id: ")) lastEventId = line.slice(4);
      }

      if (streamDone) break;
    }
  } catch (err) {
    if (err.name !== "AbortError") streamError = err;
  } finally {
    clearTimeout(heartbeatTimer);
    reader.cancel().catch(() => {});
  }

  return { done: streamDone, error: streamError, lastEventId };
}

export { streamRunWatch, WATCH_MAX_RETRIES, WATCH_RETRY_DELAY_MS, WATCH_CONNECT_TIMEOUT_MS };
